sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala

```diff
@@ -21,16 +21,160 @@ import scala.collection.mutable.ArrayBuffer
+    // we should skip it when calculating the `partitionStartIndices`.
+    val validMetrics = mapOutputStatistics.flatten
+
+    if (inputPartitionSpecs.forall(_.isEmpty)) {
+      // If all input RDDs have 0 partition, we create an empty partition for every shuffle reader.
+      if (validMetrics.isEmpty) {
+        return Seq.fill(numShuffles)(Seq(CoalescedPartitionSpec(0, 0)))
+      }
+
+      // We may have different pre-shuffle partition numbers, don't reduce shuffle partition number
+      // in that case. For example when we union fully aggregated data (data is arranged to a single
+      // partition) and a result of a SortMergeJoin (multiple partitions).
+      if (validMetrics.map(_.bytesByPartitionId.length).distinct.length > 1) {
+        return Seq.empty
+      }
+
+      val numPartitions = validMetrics.head.bytesByPartitionId.length
+      val newPartitionSpecs = coalescePartitions(
+        0, numPartitions, validMetrics, targetSize)
+      if (newPartitionSpecs.length < numPartitions) {
+        return Seq.fill(numShuffles)(newPartitionSpecs)
+      } else {
+        return Seq.empty
+      }
+    }
+
+    // Do not coalesce if any of the map output stats are missing or if not all shuffles have
+    // partition specs, which should not happen in practice.
+    if (!mapOutputStatistics.forall(_.isDefined) || !inputPartitionSpecs.forall(_.isDefined)) {
+      logWarning("Could not apply partition coalescing because of missing MapOutputStatistics " +
+        "or shuffle partition specs.")
+      return Seq.empty
+    }
+
+    // Extract the start indices of each partition spec. Give invalid index -1 to unexpected
+    // partition specs. When we reach here, it means skew join optimization has been applied.
+    val partitionIndicesSeq = inputPartitionSpecs.map(_.get.map {
+      case CoalescedPartitionSpec(start, end) if start + 1 == end => start
+      case PartialReducerPartitionSpec(reducerId, _, _, _) => reducerId
+      case _ => -1 // invalid
+    })
+
+    // There should be no unexpected partition specs and the start indices should be identical
+    // across all different shuffles.
+    assert(partitionIndicesSeq.distinct.length == 1 && partitionIndicesSeq.head.forall(_ >= 0),
+      s"Invalid shuffle partition specs: $inputPartitionSpecs")
+
+    // The indices may look like [0, 1, 2, 2, 2, 3, 4, 4, 5], and the repeated `2` and `4` mean
+    // skewed partitions.
+    val partitionIndices = partitionIndicesSeq.head
+    val newPartitionSpecsSeq = Seq.fill(numShuffles)(ArrayBuffer.empty[ShufflePartitionSpec])
+    val numPartitions = partitionIndices.length
+    var i = 0
+    var start = 0
+    while (i < numPartitions) {
+      if (i > 0 && partitionIndices(i - 1) == partitionIndices(i)) {
+        // a skew section detected, starting from partition(i - 1).
+        val repeatValue = partitionIndices(i)
+        // coalesce any partitions before partition(i - 1) and after the end of latest skew section.
```

Comment:
I think the following only does "coalesce any partitions before partition(i - 1)" but not "coalesce any partitions before partition(i - 1)"?

Reply:
`start` will be set to the end of the latest skew section.